Make it easier to add custom TNT blocks

pupnewfster opened the pull request:
Adds back an instance method variant for TNTBlock explosion, to allow mods to preform simple overrides. This fixes #5841 and also adds an AT entry to make it a lot cleaner to override the existing TNT entity without requiring each mod adding custom TNT to AT the method or have a copy of the vanilla implementation of TNTEntity#tick.

Includes a test mod that has 1/4 the fuse time and 4 times the explosion strength as vanilla TNT.

pupnewfster committed:
Make it possible for mods to easily implement custom TNT again.

pupnewfster committed:
Reduce size/complexity of patch

pupnewfster committed:
Only create one instance method and fix a small typo

pupnewfster committed:
Run update license on CustomTNTTest

pupnewfster committed:
Make custom TNT test mod spawn entity properly on client, and mimic the look of normal tnt

pupnewfster committed:
Add a small JavaDoc comment to make it more clear how the CustomTNTEntity is supposed to behave

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=6290) <br/>All committers have signed the CLA.

pupnewfster committed:
Add the new method into IForgeBlock and IForgeBlockState instead of adding it into TNTBlock. Also add an isExplosive method to both that gets checked instead of instanceof TNTBlock for purposes of calling createExplosion from FireBlock and IDispenseItemBehavior

pupnewfster commented:
I went ahead and changed it so that instead of having to extend `TNTBlock` the method for creating an explosion is in `IForgeBlock` and also added a method to check if the block is explosive so that it is possible to make blocks not extending  `TNTBlock` have the method get called when TNT normally would have its explode method called.

LexManos commented on src/main/resources/META-INF/accesstransformer.cfg:
```diff
@@ -6,6 +6,7 @@ public net.minecraft.block.Block$Properties func_200947_a(Lnet/minecraft/block/S
 public net.minecraft.block.Block$Properties func_200951_a(I)Lnet/minecraft/block/Block$Properties; # lightValue
 public net.minecraft.block.Block$Properties func_208770_d()Lnet/minecraft/block/Block$Properties; # variableOpacity
 public net.minecraft.block.Block$Properties func_222380_e()Lnet/minecraft/block/Block$Properties; # noDrops
+public net.minecraft.block.TNTBlock func_196535_a(Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/entity/LivingEntity;)V # explode
```
Keep this private, we dont want to make a function public and then ask nicely for people not to use it.
Make TNTBlock override explode and call the private method.

However, overall this PR needs to not have 'is explosive/explode' it just needs 'catchFire' and let the block manage what happens itself. Why does the outside caller care if it's explosive?

pupnewfster replied:
I will update it, but how should I then handle the IDispenseItemBehavior patch. Should I check `isFlammable` then calculate based on the source what face we are asking about or do I need to add some other check

LexManos replied:
Make catchFire return a boolean?
Also, face info would be useful. if we can figure that information out.

pupnewfster replied:
Sounds good, and I think the information can be figured out, I will look at it tomorrow.

pupnewfster committed:
Change it so that instead of adding an isExplosive and createExplosion, we just have a catchFire method

pupnewfster commented:
Ok, I updated it, and I believe it is good to go now.

pupnewfster committed:
Fix all flammable blocks being removed from dispenser lighting them on fire instead of just TNT blocks

pupnewfster committed:
Remove accidentally added import
